### PR TITLE
fix: always set default env vars for control plane

### LINKED ISSUE #63: Setting any controlplane env variable prevents default env vars from being created
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Currently default environment variables in controlplane object created set only if none of custom env variables were provided during controlplane object creation. Providing any custom env variables prevents defaults to from being created.

[controllers/controlplane_controller.go#L57](https://github.com/Kong/gateway-operator/blob/main/controllers/controlplane_controller.go#L57):

```go
debug(log, "validating ControlPlane configuration", controlplane)
if len(controlplane.Spec.Env) == 0 && len(controlplane.Spec.EnvFrom) == 0 {
	debug(log, "no ENV config found for ControlPlane resource, setting defaults", controlplane)
	setControlPlaneDefaults(&controlplane.Spec.ControlPlaneDeploymentOptions, controlplane.Namespace, nil)
	if err := r.Client.Update(ctx, controlplane); err != nil {
		if errors.IsConflict(err) {
			debug(log, "conflict found when updating ControlPlane resource, retrying", controlplane)
			return ctrl.Result{Requeue: true, RequeueAfter: requeueWithoutBackoff}, nil
		}
		return ctrl.Result{}, err
	}
	return ctrl.Result{}, nil // no need to requeue, the update will trigger.
}
```

It's not clear to me if this is intended or not.

### Expected Behavior

To be clarified.

#### Manual
```markdown
1. Set any env variable in spec of controlplane object.
2. See that defaults like `POD_NAME`, `POD_NAMESPACE`, `CONTROLLER_PUBLISH_SERVICE`, `CONTROLLER_KONG_ADMIN_URL` are not created.
```

#### Tests

1. Create a controlplane with any env var set
```go
controlplane := &operatorv1alpha1.ControlPlane{
	ObjectMeta: metav1.ObjectMeta{
		Namespace: controlplaneName.Namespace,
		Name:      controlplaneName.Name,
	},
	Spec: operatorv1alpha1.ControlPlaneSpec{
		ControlPlaneDeploymentOptions: operatorv1alpha1.ControlPlaneDeploymentOptions{
			DeploymentOptions: operatorv1alpha1.DeploymentOptions{
				Env: []corev1.EnvVar{{
					Name:  testEnvVar,
					Value: testEnvVal,
				}},
			},
			DataPlane: &nameOfDataPlane,
		},
	},
}
```

or if using gateway:
```go
gatewayConfig := &operatorv1alpha1.GatewayConfiguration{
	ObjectMeta: metav1.ObjectMeta{
		Namespace: namespace.Name,
		Name:      uuid.NewString(),
	},
	Spec: operatorv1alpha1.GatewayConfigurationSpec{
		DataPlaneDeploymentOptions: &operatorv1alpha1.DataPlaneDeploymentOptions{
			DeploymentOptions: operatorv1alpha1.DeploymentOptions{
				Env: []corev1.EnvVar{{
					Name:  testEnvVar,
					Value: testEnvVal,
				}},
			},
		},
		ControlPlaneDeploymentOptions: &operatorv1alpha1.ControlPlaneDeploymentOptions{
			DeploymentOptions: operatorv1alpha1.DeploymentOptions{
				Env: []corev1.EnvVar{{
					Name:  testEnvVar,
					Value: testEnvVal,
				}},
			},
		},
	},
}
```
2. Controlplane controller will hit the line [controllers/controlplane_controller.go#L57](https://github.com/Kong/gateway-operator/blob/main/controllers/controlplane_controller.go#L57)
3. Env variables are not set in control plane.
4. When a new deployment for control plane is created [controllers/controlplane_controller_utils.go#L69](https://github.com/Kong/gateway-operator/blob/4bf93e0c3390bdb71dba13fe0510c4f3d1607a5b/controllers/controlplane_controller_utils.go#L69), the empty ENV is copied [controllers/controlplane_controller_utils.go#L104](https://github.com/Kong/gateway-operator/blob/4bf93e0c3390bdb71dba13fe0510c4f3d1607a5b/controllers/controlplane_controller_utils.go#L104)
5. Pods created from the deployment do not have the `CONTROLLER_KONG_ADMIN_URL` set. As a result KIC will assume its default admin API endpoint.
6. Connection to admin API fail with log message `Error: could not retrieve Kong admin root: making HTTP request: Get "http://localhost:8001/": dial tcp [::1]:8001: connect: connection refused`
7. Other implications? `POD_NAME`, `POD_NAMESPACE`, `CONTROLLER_PUBLISH_SERVICE` are not set in pods too.

### Kong Ingress Controller version

```shell
N/A
```

### Kubernetes version

```shell
N/A
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by randmonkey**
**What this PR does / why we need it**:
(Changes copied from #105. Opened this because GPG signing required)
always set default environment variables for controlplane resource, fix the bug that `POD_NAME` and `POD_NAMESPACE` is not set when any `env` specified in `ControlPlane` resource.

**Which issue this PR fixes**

fixes #63 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect significant changes

### CODE REVIEW

**Review by shaneutt — DISMISSED** [review]

File: controllers/controlplane_controller_utils.go [hunk 1]
```diff
@@ -20,35 +20,62 @@ import (
...
-func setControlPlaneDefaults(spec *operatorv1alpha1.ControlPlaneDeploymentOptions, namespace string, dataplaneServiceName string, dontOverride map[string]struct{}) {
-	spec.Env = append(spec.Env, corev1.EnvVar{Name: "CONTROLLER_KONG_ADMIN_TLS_SKIP_VERIFY", Value: "true"}) // TODO: for poc only, don't release with this https://github.com/Kong/gateway-operator/issues/7
-	spec.Env = append(spec.Env, corev1.EnvVar{
-		Name: "POD_NAMESPACE",
-		ValueFrom: &corev1.EnvVarSource{
-			FieldRef: &corev1.ObjectFieldSelector{
-				APIVersion: "v1",
-				FieldPath:  "metadata.namespace",
-			},
+// setControlPlaneDefaults update the environment variables of control plane
```
Comment (pmalek) [inline controllers/controlplane_controller_utils.go:23]:
```suggestion
// setControlPlaneDefaults updates the environment variables of control plane
```

File: test/integration/controlplane_test.go [hunk 2]
```diff
@@ -156,8 +172,36 @@ func TestControlPlaneEssentials(t *testing.T) {
...
+			return fmt.Sprintf("current status of control plane: %#v", controlplane)
+		}(),
+	)
 
 	t.Log("verifying controlplane deployment has active replicas")
 	require.Eventually(t, controlPlaneHasActiveDeployment(t, controlplaneName), controlPlanetCondDeadline, controlPlanetCondTick)
+
+	// check environment variables of deployments and pods.
+	deployments := mustListControlPlaneDeployments(t, controlplane)
+	require.Len(t, deployments, 1, "must have exactly 1 deployment")
```
Comment (pmalek) [inline test/integration/controlplane_test.go:190]: Either be more specific in the error message, like: `There must be only one ControlPlane deployment` or drop this. `require`'s framework will produce a generic message similar to this.

File: test/integration/controlplane_test.go [hunk 3]
```diff
@@ -156,8 +172,36 @@ func TestControlPlaneEssentials(t *testing.T) {
...
+
+	envs := controllerContainer.Env
+	t.Log("verifying env POD_NAME comes from metadata.name")
+	podNameValueFrom := getEnvValueFromByName(envs, "POD_NAME")
+	require.True(t, reflect.DeepEqual(&corev1.EnvVarSource{
+		FieldRef: &corev1.ObjectFieldSelector{
+			APIVersion: "v1",
+			FieldPath:  "metadata.name",
+		},
+	}, podNameValueFrom))
```
Comment (pmalek) [inline test/integration/controlplane_test.go:203]: If this differs we won't see the diff, which makes the debugging harder. Can't we use `require.Equal()` or similar? Or use https://github.com/google/go-cmp when those 2 values differ to print the pretty formatted diff?

File: controllers/controlplane_controller_utils.go [hunk 4]
```diff
@@ -104,6 +131,17 @@ func envValueByName(env []corev1.EnvVar, name string) string {
 	return ""
 }
 
+// envVarSourceByName returns the VarSource of the first env var with the given name.
```
Comment (pmalek) [inline controllers/controlplane_controller_utils.go:134]: IMHO, `VarSource` in this comment should either be `ValueFrom` or `EnvVarSource`.

File: controllers/controlplane_controller_utils.go (hunk 1 above)
Comment (randmonkey) [reply controllers/controlplane_controller_utils.go]: updated.

File: controllers/controlplane_controller_utils.go (hunk 4 above)
Comment (randmonkey) [reply controllers/controlplane_controller_utils.go]: fixed.

File: test/integration/controlplane_test.go (hunk 2 above)
Comment (randmonkey) [reply test/integration/controlplane_test.go]: updated.

**Comment by jrsmroz** [comment]
@shaneutt Weren't we discussing removal of `dontOverride` all together?

**Comment by randmonkey** [comment]
> @shaneutt Weren't we discussing removal of `dontOverride` all together? It's not used at all.

I think we are still using the `dontOverride` parameter here: https://github.com/Kong/gateway-operator/blob/7ef1b7c9f6cce381e1f2b208e1cbaaf106f221ea/controllers/gateway_controller_watch.go#L188-L202

**Review by shaneutt — APPROVED** [review]

**Review by shaneutt — APPROVED** [review]

**Pull request merged**